rm-datasource/src/main/java/io/seata/rm/datasource/exec/BaseTransactionalExecutor.java

```diff
@@ -246,8 +244,30 @@ protected void prepareUndoLog(TableRecords beforeImage, TableRecords afterImage)
         }
 
         ConnectionProxy connectionProxy = statementProxy.getConnectionProxy();
-
         TableRecords lockKeyRecords = sqlRecognizer.getSQLType() == SQLType.DELETE ? beforeImage : afterImage;
+
+        List<Row> rows = lockKeyRecords.getRows();
+        if (rows.size() > 1) {
+            ConnectionContext context = connectionProxy.getContext();
+            List<SQLUndoLog> undoItems = context.getUndoItems();
+            for (SQLUndoLog undoItem : undoItems) {
+                String tableName = lockKeyRecords.getTableName();
+                TableRecords oldLockKeyRecords = sqlRecognizer.getSQLType() == SQLType.DELETE ? undoItem.getBeforeImage() : undoItem.getAfterImage();
+                String oldTableName = oldLockKeyRecords.getTableName();
+                if (oldTableName.equals(tableName)) {
+                    List<Row> oldRows = oldLockKeyRecords.getRows();
+                    for (Row oldRow : oldRows) {
+                        for (int i = rows.size() - 1; i >= 0; i--) {
+                            if (oldRow.equals(rows.get(i))){
+                                rows.remove(i);
+                            }
+                        }
+                    }
+                }
+            }
+            lockKeyRecords.setRows(rows);
+        }
+
         String lockKeys = buildLockKey(lockKeyRecords);
         connectionProxy.appendLockKey(lockKeys);
```

Comment:
Do you have any good idea when append the lock key, it will remove duplicated lock key automatically?